### PR TITLE
Use URL-safe base64 encoding for auth tokens

### PR DESCRIPTION
**Pull request by ihorserba**

### CODE REVIEW

**Comment by ihorserba** [comment]
Not well-tested yet

File: auth/php/tokenmanager.class.php [hunk 1]
```diff
@@ -4,6 +4,19 @@ class TokenManager
 {
     const TOKEN_EXPIRATION_SECONDS = 120;
 
+    /**
+     * Encode a string to base64 format.
```
Comment (vocoded) [inline auth/php/tokenmanager.class.php:8]:
```suggestion
     * Encodes a string to a URL-safe base64 format.
```

File: auth/js/tokenmanager.js [hunk 2]
```diff
@@ -3,6 +3,20 @@ const crypto = require('crypto');
 const tokenExpirationSeconds = 120;
 
 class TokenManager {
+  /**
+   * Encode data to base64 format.
```
[the remark this answers is not in the document]
Comment (ihorserba) [reply auth/js/tokenmanager.js:7]: The idea was to explain users what is going on here.
If you don't like the original comment - I propose to change as following:
```
/**
 * Encodes data to a URL-safe base64 format.
 * @param data - A string or buffer to encode
 * @returns an URL-safe base64 encoded string
 * @link https://tools.ietf.org/html/rfc4648#section-5
 */
```

File: auth/js/tokenmanager.js [hunk 3]
```diff
@@ -18,13 +32,13 @@ class TokenManager {
       exp: parseInt(new Date().getTime() / 1000, 10) + tokenExpirationSeconds
     };
 
-    const pkg = Buffer.from(JSON.stringify(header)).toString('base64') + "." + Buffer.from(JSON.stringify(payload)).toString('base64');
+    const pkg = base64UrlEncode(JSON.stringify(header)) + "." + base64UrlEncode(JSON.stringify(payload));
 
     const sign = crypto.createSign('RSA-SHA256');
     sign.update(pkg);
-    const signature = sign.sign(privateKey, 'base64');
```
[the remark this answers is not in the document]
Comment (ihorserba) [reply auth/js/tokenmanager.js:25]: We need a signed data not being encoded with base64 here.
`crypto.sign()` without second argument returns an instance of Buffer.
We use this buffer in base64UrlEncode() to encode it URL-safe.

File: auth/php/tokenmanager.class.php (hunk 1 above)
Comment (ihorserba) [reply auth/php/tokenmanager.class.php:8]: The idea was to explain users what is going on here.
If you don't like the original comment - I propose to change as following:
```
/**
 * Encodes data to a URL-safe base64 format.
 * @param data - A string or buffer to encode
 * @returns an URL-safe base64 encoded string
 * @link https://tools.ietf.org/html/rfc4648#section-5
 */
```

File: auth/js/tokenmanager.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (vocoded) [reply auth/js/tokenmanager.js:7]: My suggestion is to be clearer in the first sentence, the rest of the commentblock is fine.

File: auth/js/tokenmanager.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (vocoded) [reply auth/js/tokenmanager.js:25]: makes sense, thanks

File: auth/php/tokenmanager.class.php (hunk 1 above)
Comment (vocoded) [reply auth/php/tokenmanager.class.php:8]: see above

File: auth/js/tokenmanager.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (ihorserba) [reply auth/js/tokenmanager.js:7]: Need to explain where these '+', '/' and '=' have appeared from.
Lets add a line like:
```
Encode data to common base64.
```
so the comment is:
```
 /**
   * Encodes data to a URL-safe base64 format.
   * Encode data to common base64.
   * Replace '+' and '/' symbols with '-' and '_' respectively.
   * Remove trailing '='.
   * @param data - A string or buffer to encode
   * @returns an URL-safe base64 encoded string
   * @link https://tools.ietf.org/html/rfc4648#section-5
   */
```

File: auth/php/tokenmanager.class.php (hunk 1 above)
Comment (ihorserba) [reply auth/php/tokenmanager.class.php:8]: I would like to add one more explanation line - see in the previous comment block

File: auth/js/tokenmanager.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (vocoded) [reply auth/js/tokenmanager.js:7]: These lines read almost the same:

```
   * Encodes data to a URL-safe base64 format.
   * Encode data to common base64.
```

I understand you want to maintain the detail, so perhaps use:

```
/**
   * Encodes data to a URL-safe base64 format by taking the standard base64 output,
   * replacing '+' and '/' symbols with '-' and '_' respectively, then removing any trailing '=' symbols
   *
   * @param data - A string or buffer to encode
   * @returns The URL-safe base64 encoded form of the data
   * @link https://tools.ietf.org/html/rfc4648#section-5
   */
```

File: auth/js/tokenmanager.js [hunk 4]
```diff
@@ -3,6 +3,20 @@ const crypto = require('crypto');
...
+  /**
+   * Encode data to base64 format.
+   * Replace '+' and '/' symbols with '-' and '_' respectively.
+   * Remove trailing '='.
+   * @param data - A string or buffer to encode
+   * @returns an URL-safe base64 encoded string
+   * @link https://tools.ietf.org/html/rfc4648#section-5
+   */
+  static base64UrlEncode(data) {
+    Buffer.from
```
Comment (vocoded) [inline auth/js/tokenmanager.js:15]: not sure where this line came from, it should be removed

**Review by vocoded — CHANGES_REQUESTED** [review]
The requested change is for the invalid syntax in the JS example

File: auth/js/tokenmanager.js (hunk 4 above)
Comment (ihorserba) [reply auth/js/tokenmanager.js]: Fixed and verified js example

File: auth/js/tokenmanager.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (ihorserba) [reply auth/js/tokenmanager.js]: Replaced the comments for PHP and JS examples

**Pull request merged**